Source/Shaders/Builtin/Functions/raySphereIntersectionInterval.glsl

```diff
@@ -0,0 +1,36 @@
+/**
+ * Compute the intersection interval of a ray with a sphere.
+ *
+ * @name czm_raySphereIntersectionInterval
+ * @glslFunction
+ *
+ * @param {czm_ray} ray The ray.
+ * @param {vec3} center The center of the sphere.
+ * @param {float} radius The radius of the sphere.
+ * @return {czm_raySegment} The intersection interval of the ray with the sphere.
+ */
+czm_raySegment czm_raySphereIntersectionInterval(czm_ray ray, vec3 center, float radius)
```

Comment:
Should this function be ray-sphere intersection or ray-ellipsoid intersection?

If ray-sphere, and assuming `ray.direction` is normalized, the function can be simplified to:

```
czm_raySegment czm_rayUnitSphereIntersectionInterval(czm_ray ray)
{
    vec3 o = ray.origin;
    vec3 d = ray.direction;
    
    float b = dot(d, o);
    float c = dot(o, o) - 1.0;
    float det = b * b - c;
    
    if (det < 0.0) {
        return czm_emptyRaySegment;
    }
    
    det = sqrt(det);
    float t1 = -b - det;
    float t2 = -b + det;
    float tmin = min(t1, t2);
    float tmax = max(t1, t2);
    
    czm_raySegment result = czm_raySegment(tmin, tmax);
    return result;
}
```

And called with
```
Ray unitRay = Ray((ray.origin - center) / scale, ray.direction)
czm_raySegment intersect = czm_rayUnitSphereIntersectionInterval(unitRay);
if (intersect != czm_emptyRaySegment) {
    intersect.start *= scale;
    intersect.end *= scale;
}
```

If it's supposed to be ray-ellipsoid, the intersection code can't be simplified. But to make it work with ellipsoid the caller code would look something like below. Note how `ray.direction` is divided by `ellipsoidRadii`, putting the ray into "unit sphere space" before doing the intersection. 
```
Ray unitRay = Ray((ray.origin - center) / ellipsoidRadii, ray.direction / ellipsoidRadii)
czm_raySegment intersect = czm_rayUnitSphereIntersectionInterval(unitRay);
```

Reply:
I wonder if ellipsoid vs sphere has anything to do with the polar sky lighting issues